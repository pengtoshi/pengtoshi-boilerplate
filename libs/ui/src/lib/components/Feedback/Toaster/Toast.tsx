import clsx from "clsx";
import { toast as sonnerToast } from "sonner";
import type { UIProps } from "../../../props";
import { Icon } from "../../Icon/Icon";
import type { IconProps } from "../../Icon/Icon";
import { Button } from "../../Input/Button/Button";

export interface ToastProps extends Omit<UIProps.Div, "id"> {
  id: string | number;
  title: string;
  type?: "success" | "error" | "info";
  description?: string;
  button?: {
    label: string;
    onClick: () => void;
  };
  showIcon?: boolean;
}

const iconMap: Record<NonNullable<ToastProps["type"]>, IconProps["name"]> = {
  success: "CircleCheckFilled",
  error: "CircleErrorFilled",
  info: "CircleInfoFilled",
};

const colorMap: Record<NonNullable<ToastProps["type"]>, string> = {
  success: "text-etc-positive",
  error: "text-etc-negative",
  info: "text-gray-50",
};

export const Toast = ({ id, title, type = "info", description, button, showIcon = true }: ToastProps) => {
  return (
    <div className="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3">
      <div className="flex items-center justify-start gap-2">
        {showIcon && <Icon size={20} name={iconMap[type]} className={colorMap[type]} />}
        <div className="flex flex-col gap-1">
          <p className={clsx("text-14/body/emp", colorMap[type])}>{title}</p>
          {description && <p className="text-12/body text-gray-400">{description}</p>}
        </div>
      </div>
      {!!button && (
        <Button
          size="small"
          className="!w-fit !bg-gray-50 !text-gray-950"
          onClick={() => {
            button.onClick();
            sonnerToast.dismiss(id);
          }}
        >
          {button.label}
        </Button>
      )}
    </div>
  );
};
