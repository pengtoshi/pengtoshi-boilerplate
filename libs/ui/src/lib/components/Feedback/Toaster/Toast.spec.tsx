import { composeStories } from "@storybook/react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as stories from "./Toast.stories";

describe("Toast", () => {
  Object.entries(composeStories(stories)).forEach(([, Story]) => {
    it("should render", () => {
      const { baseElement } = render(<Story />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
