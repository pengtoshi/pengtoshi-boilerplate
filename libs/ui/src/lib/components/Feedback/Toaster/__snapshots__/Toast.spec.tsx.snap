// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render 1`] = `
<body>
  <div>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
    <div
      class="flex w-[360px] flex-col gap-4"
    >
      <div
        class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
      >
        <div
          class="flex items-center justify-start gap-2"
        >
          <div
            class="text-gray-50"
            height="20"
            width="20"
          />
          <div
            class="flex flex-col gap-1"
          >
            <p
              class="text-14/body/emp text-gray-50"
            >
              Example Toast
            </p>
            <p
              class="text-12/body text-gray-400"
            >
              This is toast description.
            </p>
          </div>
        </div>
      </div>
      <button
        class="group inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 !text-gray-50 disabled:bg-gray-150 disabled:text-gray-350 h-10 px-5 text-14/button gap-1.5 relative overflow-hidden w-[160px]"
      >
        <div
          class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
        />
        <span
          class="relative z-10"
        >
          Show Toast
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Toast should render 2`] = `
<body>
  <div>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
    <div
      class="flex flex-col items-start gap-4"
    >
      <div
        class="flex items-center gap-6"
      >
        <span
          class="w-[100px] text-14/body text-gray-600"
        >
          default
        </span>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
      </div>
      <div
        class="flex items-center gap-6"
      >
        <span
          class="w-[100px] text-14/body text-gray-600"
        >
          success
        </span>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-positive"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-positive"
              >
                Example Toast
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-positive"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-positive"
              >
                Example Toast
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-positive"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-positive"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-positive"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-positive"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
      </div>
      <div
        class="flex items-center gap-6"
      >
        <span
          class="w-[100px] text-14/body text-gray-600"
        >
          error
        </span>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-negative"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-negative"
              >
                Example Toast
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-negative"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-negative"
              >
                Example Toast
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-negative"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-negative"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-etc-negative"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-etc-negative"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
      </div>
      <div
        class="flex items-center gap-6"
      >
        <span
          class="w-[100px] text-14/body text-gray-600"
        >
          info
        </span>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex w-[360px] items-center justify-between gap-3 rounded-md bg-gray-950 px-4 py-3"
        >
          <div
            class="flex items-center justify-start gap-2"
          >
            <div
              class="text-gray-50"
              height="20"
              width="20"
            />
            <div
              class="flex flex-col gap-1"
            >
              <p
                class="text-14/body/emp text-gray-50"
              >
                Example Toast
              </p>
              <p
                class="text-12/body text-gray-400"
              >
                This is toast description.
              </p>
            </div>
          </div>
          <button
            class="group w-fit inline-flex items-center justify-center whitespace-nowrap rounded-md transition-all duration-300 disabled:pointer-events-none bg-primary-500 disabled:bg-gray-150 disabled:text-gray-350 h-8 px-4 text-14/button gap-1 relative overflow-hidden !w-fit !bg-gray-50 !text-gray-950"
          >
            <div
              class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12]"
            />
            <span
              class="relative z-10"
            >
              Click me
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
